aws_topology/stackstate_checks/aws_topology/aws_topology.py

```diff
@@ -0,0 +1,158 @@
+
+from .resources import ResourceRegistry
+
+memory_data = {}  # name -> arn for cloudformation
+
+DEFAULT_BOTO3_RETRIES_COUNT = 50
+
+DEFAULT_BOTO3_CONFIG = Config(
+    retries=dict(
+        max_attempts=DEFAULT_BOTO3_RETRIES_COUNT
+    )
+)
+
+DEFAULT_COLLECTION_INTERVAL = 60
+
+
+class State(Model):
+    code = StringType(required=True)
+
+
+class InstanceInfo(Model):
+    region = StringType(required=True)
+    aws_access_key_id = StringType(required=True)
+    aws_secret_access_key = StringType(required=True)
+    role_arn = StringType(required=True)
+    account_id = StringType(required=True)
+    tags = ListType(StringType, default=[])
+    arns = DictType(StringType, default={})
+    apis_to_run = ListType(StringType)
+
+
+class AwsTopologyCheck(AgentCheck):
+    """Collects AWS Topology and sends them to STS."""
+    INSTANCE_TYPE = 'aws'  # TODO should we add _topology?
+    SERVICE_CHECK_CONNECT_NAME = 'aws_topology.can_connect'
+    SERVICE_CHECK_EXECUTE_NAME = 'aws_topology.can_execute'
+    INSTANCE_SCHEMA = InstanceInfo
+
+    def get_registry(self):
+        return ResourceRegistry.get_registry()
+
+    def get_instance_key(self, instance_info):
+        return TopologyInstance(self.INSTANCE_TYPE, str(instance_info.account_id))
+
+    def check(self, instance_info):
+        try:
+            aws_client = AwsClient(instance_info, self.init_config)
+            instance_info.region = aws_client.region
+            account_id = aws_client.get_account_id()
+            if not account_id == instance_info.account_id:
+                raise Exception(
+                    "AWS caller identity does not return correct account_id. %s was returned, but %s was expected." %
+                    (account_id, instance_info.account_id)
+                )
+            self.service_check(self.SERVICE_CHECK_CONNECT_NAME, AgentCheck.OK, tags=instance_info.tags)
+        except Exception as e:
+            msg = 'AWS connection failed: {}'.format(e)
+            self.log.error(msg)
+            self.service_check(
+                self.SERVICE_CHECK_CONNECT_NAME,
+                AgentCheck.CRITICAL,
+                message=msg,
+                tags=instance_info.tags
+            )
+            return
+
+        try:
+            self._get_topology(instance_info, aws_client)
+            self.service_check(self.SERVICE_CHECK_EXECUTE_NAME, AgentCheck.OK, tags=instance_info.tags)
+        except Exception as e:
+            msg = 'AWS topology collection failed: {}'.format(e)
+            self.log.error(msg)
+            self.service_check(
+                self.SERVICE_CHECK_EXECUTE_NAME,
+                AgentCheck.CRITICAL,
+                message=msg,
+                tags=instance_info.tags
+            )
+
+    def _get_topology(self, instance_info, aws_client):
+        """Gets AWS Topology returns them in Agent format."""
+        self.start_snapshot()
+
+        errors = []
+        registry = self.get_registry()
```

Comment:
I like the idea of having some kind of registry where all of the "collectors" are stored, but I don't like this part below that has to make decisions about what needs to be run. This filtering / decision-making should be done in the collector itself. I feel like the separation between the agent check, the AWS client and the collector is not hard enough and so the logic of how it all fits together filters through the cracks of the separation.

Reply:
Need a talk about this one :-)